App/Logic/DataUploadLogic.swift

```diff
@@ -232,22 +266,181 @@ extension Logic.DataUpload {
+      let state = context.getState()
+
+      let now = context.dependencies.now()
+
+      guard now > state.ingestion.dummyTrafficOpportunityWindow.windowEnd else {
+        // Window not started or finished yet.
+        return
+      }
+
+      try context.awaitDispatch(UpdateDummyTrafficOpportunityWindow())
+    }
+  }
+
+  /// Updates the dummy ingestion traffic opportunity window taking the parameters from the Configuration and the RNGs.
+  struct UpdateDummyTrafficOpportunityWindow: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      let dummyTrafficStochasticDelay = context.dependencies.exponentialDistributionGenerator
+        .exponentialRandom(with: state.configuration.dummyIngestionMeanStochasticDelay)
+
+      try context.awaitDispatch(
+        SetDummyTrafficOpportunityWindow(
+          dummyTrafficStochasticDelay: dummyTrafficStochasticDelay,
+          dummyTrafficOpportunityWindowDuration: state.configuration.dummyIngestionWindowDuration,
+          now: context.dependencies.now()
+        )
+      )
+    }
+  }
+
+  /// Schedules a sequence of dummy ingestion requests for some random point in the future.
+  /// -seeAlso: https://github.com/immuni-app/immuni-documentation/blob/master/Traffic%20Analysis%20Mitigation.md
+  struct ScheduleDummyIngestionSequenceIfNecessary: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      guard !state.ingestion.dummyTrafficSequenceScheduledInSession else {
+        // There is already another sequence schedule for this session.
+        return
+      }
+
+      // Reset the flag in case this is a new session
+      try context.awaitDispatch(SetDummyTrafficSequenceCancelled(value: false))
+
+      guard state.ingestion.dummyTrafficOpportunityWindow.contains(context.dependencies.now()) else {
+        // Not within the opportunity window. Nothing to do.
+        return
+      }
+
+      // Prevent other ingestion sequences from being scheduled in this session
+      try context.awaitDispatch(SetDummyIngestionSequenceScheduledForThisSession(value: true))
+
+      let startDelay = context.dependencies.exponentialDistributionGenerator
+        .exponentialRandom(with: state.configuration.dummyIngestionAverageStartUpDelay)
+
+      // Dispatch a delayed action and return
+      context.dispatch(StartIngestionSequenceIfNotCancelled().deferred(of: startDelay))
+    }
+  }
+
+  /// Attempts a simulated sequence of ingestion requests, unless it was cancelled for outside in the meanwhile.
+  /// -seeAlso: https://github.com/immuni-app/immuni-documentation/blob/master/Traffic%20Analysis%20Mitigation.md
+  struct StartIngestionSequenceIfNotCancelled: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      var executions = 0
+      while true {
```

Comment:
can we use a bool variable? just for our piece of mind

Reply:
Honestly it wouldn't increase my peace of mind. You could still never set that to false.